sim,mem,cpu,arch-arm: Add support for cache PMU events

arichardson opened the pull request:
See individual commits for details.

This was rather tricky to figure out - I noticed I could not check the dynamic type of the connected probes since we were calling virtual functions on partially constructed objects.

arichardson committed:
sim: Ensure probe pointers are connected to compatible listeners

I was trying to add PMU events for caches and noticed that we were
calling the notify() function with a nonsense increment since cache
events pass a CacheAccessProbeArg& to notify() and the RegularProbe in
the Arm PMU expects a uint64_t&. The result was that the counters was
incremented by the value of the PacketPtr member in the
CacheAccessProbeArg instead of incrementing by one. To avoid similar
difficult to debug issues, use a dynamic_cast<> in addListener().
This assertion is possible now that we no longer pass partially
constructed objects to this function.

Change-Id: I4aba443b46df5fbbe41d5fb6f9b2fe424d9f806e

giactra commented:
Wow this looks like an amazing contribution.
As it is touching non arch-arm code as well, can I ask you to amend the PR title? So that other people can chime in if interested on reviewing this

powerjg commented:
@studyztp can you review this change with respect to the probe changes? You just spent a bunch of time writing the documentation for probes so I think you would be a great person to take a look at this part of the change.

giactra reviewed (commented):
Hi @arichardson 

I didn't do a full review of the PR. It generally looks fine, however I am a bit concerned we are adding extra abstractions which might not be fully necessary

The concept of Probe and Listener is quite straightforward per se. You are adding the ProbeConnection abstraction and as far as I can see it is simply adding an indirection to a listener. Do we really need that?

```
class ProbeConnection
{
  public:
    ProbeConnection(ProbeManager* _manager,
               std::unique_ptr<ProbeListener> _listener);
    ~ProbeConnection();
    ProbeConnection(const ProbeConnection& other) = delete;

 protected:
    std::unique_ptr<ProbeListener> listener;
    ProbeManager* manager;
};
```

You solve the partially construct problem by splitting the Listener construction and Listener connection in two steps within the connect method in the probe manager

    template<typename Listener, typename... Args>
    ProbeConnectionPtr connect(Args&&... args)
    {
        return std::make_unique<ProbeConnection>(
            this, std::make_unique<Listener>(std::forward<Args>(args)...));
    }

Can't we just do this without defining an intermediate object? The ProbeConnector constructor is anyway calling the manager addListener method, so we have

ProbeManager::connect -> calls -> ProbeConnection::ProbeConnextion-> calls -> ProbeManager::addListener....

IMHO we could call addListener within the connect method straight away and return the listener ptr directly... Please let me know your thought and if there's something I am missing

studyztp commented:
@arichardson, this is a great work! I looked over all the changes. Most of them make sense to me. I just want to discuss a few things. Sorry for the out-of-order comments. I did not realize that my review would be published immediately :)

arichardson commented:
> Hi @arichardson
> 
> I didn't do a full review of the PR. It generally looks fine, however I am a bit concerned we are adding extra abstractions which might not be fully necessary
> 
> The concept of Probe and Listener is quite straightforward per se. You are adding the ProbeConnection abstraction and as far as I can see it is simply adding an indirection to a listener. Do we really need that?
> 
> ```
> class ProbeConnection
> {
>   public:
>     ProbeConnection(ProbeManager* _manager,
>                std::unique_ptr<ProbeListener> _listener);
>     ~ProbeConnection();
>     ProbeConnection(const ProbeConnection& other) = delete;
> 
>  protected:
>     std::unique_ptr<ProbeListener> listener;
>     ProbeManager* manager;
> };
> ```
> 
> You solve the partially construct problem by splitting the Listener construction and Listener connection in two steps within the connect method in the probe manager
> 
> ```
> template<typename Listener, typename... Args>
> ProbeConnectionPtr connect(Args&&... args)
> {
>     return std::make_unique<ProbeConnection>(
>         this, std::make_unique<Listener>(std::forward<Args>(args)...));
> }
> ```
> 
> Can't we just do this without defining an intermediate object? The ProbeConnector constructor is anyway calling the manager addListener method, so we have
> 
> ProbeManager::connect -> calls -> ProbeConnection::ProbeConnextion-> calls -> ProbeManager::addListener....
> 
> IMHO we could call addListener within the connect method straight away and return the listener ptr directly... Please let me know your thought and if there's something I am missing

The reason I added the intermediate object is that I didn't see an easy solution to avoid the virtual function call from within the constructor without this indirection. It basically guarantees that we have fully constructed the listener before calling addListener(). Otherwise we end up passing a partially constructed object and that is dangerous.

There might be a cleaner solution, but this seemed like a reasonable approach to me.

arichardson commented:
> @arichardson, this is a great work! I looked over all the changes. Most of them make sense to me. I just want to discuss a few things. Sorry for the out-of-order comments. I did not realize that my review would be published immediately :)

Hi, I believe your github comments were not published. Probably need to click on finish review?

arichardson commented:
> > Hi @arichardson
> > I didn't do a full review of the PR. It generally looks fine, however I am a bit concerned we are adding extra abstractions which might not be fully necessary
> > The concept of Probe and Listener is quite straightforward per se. You are adding the ProbeConnection abstraction and as far as I can see it is simply adding an indirection to a listener. Do we really need that?
> > ```
> > class ProbeConnection
> > {
> >   public:
> >     ProbeConnection(ProbeManager* _manager,
> >                std::unique_ptr<ProbeListener> _listener);
> >     ~ProbeConnection();
> >     ProbeConnection(const ProbeConnection& other) = delete;
> > 
> >  protected:
> >     std::unique_ptr<ProbeListener> listener;
> >     ProbeManager* manager;
> > };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You solve the partially construct problem by splitting the Listener construction and Listener connection in two steps within the connect method in the probe manager
> > ```
> > template<typename Listener, typename... Args>
> > ProbeConnectionPtr connect(Args&&... args)
> > {
> >     return std::make_unique<ProbeConnection>(
> >         this, std::make_unique<Listener>(std::forward<Args>(args)...));
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Can't we just do this without defining an intermediate object? The ProbeConnector constructor is anyway calling the manager addListener method, so we have
> > ProbeManager::connect -> calls -> ProbeConnection::ProbeConnextion-> calls -> ProbeManager::addListener....
> > IMHO we could call addListener within the connect method straight away and return the listener ptr directly... Please let me know your thought and if there's something I am missing
> 
> The reason I added the intermediate object is that I didn't see an easy solution to avoid the virtual function call from within the constructor without this indirection. It basically guarantees that we have fully constructed the listener before calling addListener(). Otherwise we end up passing a partially constructed object and that is dangerous.
> 
> There might be a cleaner solution, but this seemed like a reasonable approach to me.

Essentially, this ProbeConnection class is just a std::unique_ptr<ProbeListener>, but instead of causing undefined behaviour by calling {add/remove}Listener in the {con/de}structor it calls those methods while the object is in a fully constructed state.

arichardson commented:
> > > Hi @arichardson
> > > I didn't do a full review of the PR. It generally looks fine, however I am a bit concerned we are adding extra abstractions which might not be fully necessary
> > > The concept of Probe and Listener is quite straightforward per se. You are adding the ProbeConnection abstraction and as far as I can see it is simply adding an indirection to a listener. Do we really need that?
> > > ```
> > > class ProbeConnection
> > > {
> > >   public:
> > >     ProbeConnection(ProbeManager* _manager,
> > >                std::unique_ptr<ProbeListener> _listener);
> > >     ~ProbeConnection();
> > >     ProbeConnection(const ProbeConnection& other) = delete;
> > > 
> > >  protected:
> > >     std::unique_ptr<ProbeListener> listener;
> > >     ProbeManager* manager;
> > > };
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > You solve the partially construct problem by splitting the Listener construction and Listener connection in two steps within the connect method in the probe manager
> > > ```
> > > template<typename Listener, typename... Args>
> > > ProbeConnectionPtr connect(Args&&... args)
> > > {
> > >     return std::make_unique<ProbeConnection>(
> > >         this, std::make_unique<Listener>(std::forward<Args>(args)...));
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Can't we just do this without defining an intermediate object? The ProbeConnector constructor is anyway calling the manager addListener method, so we have
> > > ProbeManager::connect -> calls -> ProbeConnection::ProbeConnextion-> calls -> ProbeManager::addListener....
> > > IMHO we could call addListener within the connect method straight away and return the listener ptr directly... Please let me know your thought and if there's something I am missing
> > 
> > 
> > The reason I added the intermediate object is that I didn't see an easy solution to avoid the virtual function call from within the constructor without this indirection. It basically guarantees that we have fully constructed the listener before calling addListener(). Otherwise we end up passing a partially constructed object and that is dangerous.
> > There might be a cleaner solution, but this seemed like a reasonable approach to me.
> 

Essentially, this ProbeConnection class is just a std::unique_ptr, but instead of causing undefined behaviour by calling {add/remove}Listener in the {con/de}structor it calls those methods while the object is in a fully constructed state. And I agree, we could call addListener in the connect method, but I'm not sure calling removeListener with a partially destructed object is actually legal C++.

giactra commented:
> > Hi @arichardson
> > I didn't do a full review of the PR. It generally looks fine, however I am a bit concerned we are adding extra abstractions which might not be fully necessary
> > The concept of Probe and Listener is quite straightforward per se. You are adding the ProbeConnection abstraction and as far as I can see it is simply adding an indirection to a listener. Do we really need that?
> > ```
> > class ProbeConnection
> > {
> >   public:
> >     ProbeConnection(ProbeManager* _manager,
> >                std::unique_ptr<ProbeListener> _listener);
> >     ~ProbeConnection();
> >     ProbeConnection(const ProbeConnection& other) = delete;
> > 
> >  protected:
> >     std::unique_ptr<ProbeListener> listener;
> >     ProbeManager* manager;
> > };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You solve the partially construct problem by splitting the Listener construction and Listener connection in two steps within the connect method in the probe manager
> > ```
> > template<typename Listener, typename... Args>
> > ProbeConnectionPtr connect(Args&&... args)
> > {
> >     return std::make_unique<ProbeConnection>(
> >         this, std::make_unique<Listener>(std::forward<Args>(args)...));
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Can't we just do this without defining an intermediate object? The ProbeConnector constructor is anyway calling the manager addListener method, so we have
> > ProbeManager::connect -> calls -> ProbeConnection::ProbeConnextion-> calls -> ProbeManager::addListener....
> > IMHO we could call addListener within the connect method straight away and return the listener ptr directly... Please let me know your thought and if there's something I am missing
> 
> The reason I added the intermediate object is that I didn't see an easy solution to avoid the virtual function call from within the constructor without this indirection. It basically guarantees that we have fully constructed the listener before calling addListener(). Otherwise we end up passing a partially constructed object and that is dangerous.
> 
> There might be a cleaner solution, but this seemed like a reasonable approach to me.

Yes it is more than reasonable and it is also clean. I am just being concerned about introducing new abstractions. Also lots of members are still called listeners even if they hold a vector of connection pointers. What I was proposing was something like the following

    template<typename Listener, typename... Args>
    std::unique_ptr<Listener> connect(Args&&... args)
    {
        auto listener = std::make_unique<Listener>(std::forward<Args>(args)...));
        addListener(listener->getName(), *listener);
        return std::move(listener);
    }

(Note we might want to rename connect method in that case). I am not sure my implementation is better than yours. Using ConnectionPtr might have some other advantages so I am not going to push strongly against ProbeConnection. I am more interested on hearing your thoughts over my semplification

giactra commented on src/arch/arm/pmu.hh:
```diff
@@ -358,28 +359,40 @@ class PMU : public SimObject, public ArmISA::BaseISADevice
+                : ProbeListenerArgBase(std::move(name)), parentEvent(parent) {}
+            CacheProbe() = delete;
+
+            void notify(const CacheAccessProbeArg &val) override {
```
Nit: this shoul be:

```
void
notify(const CacheAccessProbeArg &val) override
{
```

arichardson replied:
Ah, sorry I contribute to too many different projects to always remember the right coding style... I wish we could get #362 landed. Will fix the style issues.

giactra commented on src/sim/probe/probe.cc:
```diff
@@ -61,61 +61,39 @@ ProbeListenerObject::ProbeListenerObject(
+    if (ProbePoint *p = getProbePoint(point_name)) {
+        p->addListener(&listener);
+        return true;
+    } else {
```
It has to be noted that in the previous implementation the loop was iterating over all probe points and the listener was being registered every time there was a match. In the new implementation you only register the listener once and return afterwards.
This is ok as long as probe names are unique. The question is, can you have multiple probes with the same name?

arichardson replied:
Oh good catch! I didn't even consider more than one. Will restore the old logic in case anyone relies on this.

giactra commented on src/sim/probe/probe.cc:
```diff
@@ -61,61 +61,39 @@ ProbeListenerObject::ProbeListenerObject(
+    if (ProbePoint *p = getProbePoint(point_name)) {
+        p->removeListener(&listener);
+        return true;
+    } else {
```
Ditto

Harshil2107 commented:
Hi @arichardson,

Thank you for the contribution, could you please update us on the status of this PR?

arichardson commented:
> Hi @arichardson,
> 
> Thank you for the contribution, could you please update us on the status of this PR?

Sorry I've been caught up on other projects for the past month, will try to address the comments shortly

erin-le commented:
@arichardson Hello, could you provide another update on what the status of this PR is? Thank you!

giactra commented:
@odanrc

erin-le commented:
@arichardson This PR is being closed for inactivity, but can be reopened if you would like to continue working on it. Thank you!

arichardson committed:
sim,mem,cpu,arch-arm: Avoid virtual calls on partially constructed probes

Calling manager->addListener() in the ProbeListener constructor means
we are passing a partially constructed object to the addListener()
function. I noticed this because I tried to add a dynamic_cast<> inside
addListener() to assert the function signatures match and it was not
working as expected since the object passed in has a vtable pointer
that points to the base class rather than the derived class during the
partially constructed state.

This introduces a new ProbeListenerPtr unique_ptr specialization with
a custom deleter that takes care of calling removeListener() when the
listener is destroyed. It also adds a new connect() method to
ProbeManager to ensure that the addListener() method is only called
once the ProbeListener is fully constructed.

Change-Id: Ic9b2f22c148e8bd66fd3d04c7a32ac6cb2b041a5

arichardson committed:
sim,arch-arm: Support cache events in the Arm PMU

This adds a new type of Probe that increments by one for cache events
and registers the standard Arm L1I,L1D and L2 events in ArmPMU.
I confirmed that these values increment as expected using a simple
test program that reads the PMUs.

Change-Id: I76ecf7f6b68c5c997d0ed531a77b99dd1692899d

arichardson committed:
sim: Use std::string_view instead of std::string

No need to copy the string here.

Change-Id: I55a6cdbb1b35e8e840bbc3fec1df21bf869d8a1a

arichardson commented:
Rebased and hopefully fixed all issues. @erin-le @giactra could you re-open it?

I changed the custom wrapper class to a unique_ptr with a custom deleter which should make it much simpler now.

erin-le commented:
Hello @arichardson, when I try to reopen the PR, I get a message saying that the branch was force-pushed or recreated. I'm not sure this will work, but could you try the steps listed [here](https://gist.github.com/robertpainsi/2c42c15f1ce6dab03a0675348edd4e2c) and see if that works to reopen it? Sorry for the inconvenience.

arichardson commented:
@erin-le I have pushed the old version again but I don't have permissions to reopen the PR. If you are able to re-open it can then push the new version. If that doesn't work I'll just go ahead and create a new PR.

giactra reviewed (approved):
Thanks for your changes!

heshpdx commented:
@powerjg We were hitting a sanitizer issue in the SPEC codebase (which is based off of v22.1). I found this PR which is a likely fix. I manually took the patches from here and put them my codebase (along with #2291). I reran sanitizer and the issue went away. Just thought I would share with the community. Thank you!

```
gem5/sim/probe/probe.hh:293:33: runtime error: downcast of address 0x10d1b3be2120 which does not point to an object of type 'ProbeListenerArgBase'
0x10d1b3be2120: note: object is of type 'gem5::ProbeListener'
 00 00 00 00  00 e1 c1 0a 00 00 00 00  d0 5f 49 b6 b1 10 00 00  40 21 be b3 d1 10 00 00  04 00 00 00
              ^~~~~~~~~~~~~~~~~~~~~~~
              vptr for 'gem5::ProbeListener'
    #0 0x00000231aab0 in gem5::ProbePointArg<gem5::Packet*>::addListener(gem5::ProbeListener*) gem5/sim/probe/probe.hh:293
    #1 0x000008f192c4 in gem5::ProbeManager::addListener(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, gem5::ProbeListener&) gem5/sim/probe/probe.cc:91
    #2 0x000008f1c3a9 in gem5::ProbeListener::ProbeListener(gem5::ProbeManager*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) gem5/sim/probe/probe.cc:75
    #3 0x000003e7486a in gem5::ProbeListenerArgBase<gem5::Packet*>::ProbeListenerArgBase(gem5::ProbeManager*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) gem5/sim/probe/probe.hh:215
    #4 0x000003e7486a in gem5::prefetch::Base::PrefetchListener::PrefetchListener(gem5::prefetch::Base&, gem5::ProbeManager*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, bool) gem5/mem/cache/prefetch/base.hh:81
    #5 0x000003e73b36 in gem5::prefetch::Base::regProbeListeners() gem5/mem/cache/prefetch/base.cc:286
    #6 0x000007e70b75 in pybind11::cpp_function::cpp_function<void, gem5::SimObject, , pybind11::name, pybind11::is_method, pybind11::sibling>(void (gem5::SimObject::*)(), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&)::{lambda(gem5::SimObject*)#1}::operator()(gem5::SimObject*) const ext/pybind11/include/pybind11/pybind11.h:109
```

arichardson commented:
> @powerjg We were hitting a sanitizer issue in the SPEC codebase (which is based off of v22.1). I found this PR which is a likely fix. I manually took the patches from here and put them my codebase (along with #2291). I reran sanitizer and the issue went away. Just thought I would share with the community. Thank you!
> 
> ```
> gem5/sim/probe/probe.hh:293:33: runtime error: downcast of address 0x10d1b3be2120 which does not point to an object of type 'ProbeListenerArgBase'
> 0x10d1b3be2120: note: object is of type 'gem5::ProbeListener'
>  00 00 00 00  00 e1 c1 0a 00 00 00 00  d0 5f 49 b6 b1 10 00 00  40 21 be b3 d1 10 00 00  04 00 00 00
>               ^~~~~~~~~~~~~~~~~~~~~~~
>               vptr for 'gem5::ProbeListener'
>     #0 0x00000231aab0 in gem5::ProbePointArg<gem5::Packet*>::addListener(gem5::ProbeListener*) gem5/sim/probe/probe.hh:293
>     #1 0x000008f192c4 in gem5::ProbeManager::addListener(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, gem5::ProbeListener&) gem5/sim/probe/probe.cc:91
>     #2 0x000008f1c3a9 in gem5::ProbeListener::ProbeListener(gem5::ProbeManager*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) gem5/sim/probe/probe.cc:75
>     #3 0x000003e7486a in gem5::ProbeListenerArgBase<gem5::Packet*>::ProbeListenerArgBase(gem5::ProbeManager*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) gem5/sim/probe/probe.hh:215
>     #4 0x000003e7486a in gem5::prefetch::Base::PrefetchListener::PrefetchListener(gem5::prefetch::Base&, gem5::ProbeManager*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, bool) gem5/mem/cache/prefetch/base.hh:81
>     #5 0x000003e73b36 in gem5::prefetch::Base::regProbeListeners() gem5/mem/cache/prefetch/base.cc:286
>     #6 0x000007e70b75 in pybind11::cpp_function::cpp_function<void, gem5::SimObject, , pybind11::name, pybind11::is_method, pybind11::sibling>(void (gem5::SimObject::*)(), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&)::{lambda(gem5::SimObject*)#1}::operator()(gem5::SimObject*) const ext/pybind11/include/pybind11/pybind11.h:109
> ```

Yeah this looks very much like the issue this patch fixes. Ubsan is detecting the use of a partially constructed object. The vtable pointer only gets updated once the 
ProbeListenerArgBase constructor has completed calling all base class constructors.
